### PR TITLE
tkt-83924: Normalize/Validate Crontab Commands

### PR DESCRIPTION
**Pull request by sonicaj**
This PR aims to validate cron commands making sure they donot exceed 1000 characters as FreeBSD enforces a hard limit. It also normalizes how we write commands in crontab ensuring that we donot insert unnecessary spaces which consume 1000characters buffer.

Ticket: #83924

### CODE REVIEW

File: src/middlewared/middlewared/plugins/cron.py [hunk 1]
```diff
@@ -41,14 +48,38 @@ def cron_extend(self, data):
...
+        else:
+            allowed_length = 1000 - len(' '.join(
+                filter(
+                    bool, (
+                        f'PATH="{await self.cronjob_path()}"',
+                        '> /dev/null' if data['stdout'] else '',
+                        '2> /dev/null' if data['stderr'] else ''
+                    )
+                )
+            ))
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/cron.py]: Seems we are repeating ourselves here, arent we?
Why not make a private method to get the full command, including PATH and use it in cron generation?

Then `cronjob_path` can also go away as i dont see that being useful for public API.

File: src/middlewared/middlewared/plugins/cron.py [hunk 2]
```diff
@@ -41,14 +48,38 @@ def cron_extend(self, data):
...
                 )
 
+        command = data.get('command')
+        if not command:
+            verrors.add(
+                f'{schema}.command',
+                'Please specify a command for cronjob task.'
+            )
+        else:
+            allowed_length = 1000 - len(' '.join(
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/cron.py]: Ah, also a comment about why the magical 1000, please. The next developer working here will appreciate ;)

File: src/middlewared/middlewared/plugins/cron.py [hunk 3]
```diff
@@ -16,6 +16,20 @@ def cron_extend(self, data):
         Cron.convert_db_format_to_schedule(data)
         return data
 
+    @private
+    async def construct_cron_command(self, schedule, user, command, stdout=True, stderr=True):
+        return list(
+            filter(
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/cron.py]: What is this filter for?

Comment (sonicaj) [reply src/middlewared/middlewared/plugins/cron.py]: So when we join this list on space, if `stdout` is `''`, it adds an extra space in the final command. This filter flushes `stdout/stderr` if they are emtpy

File: src/middlewared/middlewared/plugins/cron.py [hunk 4]
```diff
@@ -41,14 +55,38 @@ def cron_extend(self, data):
...
+                'Please specify a command for cronjob task.'
+            )
+        else:
+            crontab_cmd = (await self.construct_cron_command(
+                data['schedule'], user, command, data['stdout'], data['stderr']
+            ))[6:]
+            command = crontab_cmd.pop(1)
+
+            # cron(8) when reads an entry from a crontab, it keeps a buffer of 1000 characters and anything more then
+            # that is truncated. So we validate that the user supplied command is not of more then 1000 characters.
```
Comment (Mrt134) [inline src/middlewared/middlewared/plugins/cron.py]:
```suggestion
            # that is truncated. We validate that the user supplied command is not more than 1000 characters.
```

File: src/middlewared/middlewared/plugins/cron.py [hunk 5]
```diff
@@ -41,14 +55,38 @@ def cron_extend(self, data):
...
+                f'{schema}.command',
+                'Please specify a command for cronjob task.'
+            )
+        else:
+            crontab_cmd = (await self.construct_cron_command(
+                data['schedule'], user, command, data['stdout'], data['stderr']
+            ))[6:]
+            command = crontab_cmd.pop(1)
+
+            # cron(8) when reads an entry from a crontab, it keeps a buffer of 1000 characters and anything more then
```
Comment (Mrt134) [inline src/middlewared/middlewared/plugins/cron.py]:
```suggestion
            # When cron(8) reads an entry from a crontab, it keeps a buffer of 1000 characters and anything more then
```

File: src/middlewared/middlewared/plugins/cron.py [hunk 6]
```diff
@@ -41,14 +55,38 @@ def cron_extend(self, data):
...
+            # that is truncated. So we validate that the user supplied command is not of more then 1000 characters.
+            allowed_length = 1000 - len(' '.join(crontab_cmd))
+
+            if len(command) > allowed_length:
+                verrors.add(
+                    f'{schema}.command',
+                    f'Command must be less then or equal to {allowed_length} characters. '
+                    'It should be noted that in the command provided, all newline characters are removed and '
```
Comment (Mrt134) [inline src/middlewared/middlewared/plugins/cron.py]:
```suggestion
                    'Any newline or "%" characters in the command are automatically removed or escaped with '
```

File: src/middlewared/middlewared/plugins/cron.py [hunk 7]
```diff
@@ -41,14 +55,38 @@ def cron_extend(self, data):
...
+                    '"%" characters are escaped with backslash.'
```
Comment (Mrt134) [inline src/middlewared/middlewared/plugins/cron.py]:
```suggestion
                    'backslash.'
```

File: src/middlewared/middlewared/plugins/cron.py [hunk 8]
```diff
@@ -41,14 +55,38 @@ def cron_extend(self, data):
...
+            command = crontab_cmd.pop(1)
+
+            # cron(8) when reads an entry from a crontab, it keeps a buffer of 1000 characters and anything more then
+            # that is truncated. So we validate that the user supplied command is not of more then 1000 characters.
+            allowed_length = 1000 - len(' '.join(crontab_cmd))
+
+            if len(command) > allowed_length:
+                verrors.add(
+                    f'{schema}.command',
+                    f'Command must be less then or equal to {allowed_length} characters. '
```
Comment (Mrt134) [inline src/middlewared/middlewared/plugins/cron.py]:
```suggestion
                    f'Command must be less than or equal to {allowed_length} characters. '
```

File: src/middlewared/middlewared/plugins/cron.py (hunk 6 above)
Comment (sonicaj) [reply src/middlewared/middlewared/plugins/cron.py]: @Mrt134 newline characters are not escaped with a backslash ;) How about we make that explicit here ? Or here implies that both cases can happen ( removed/escaped ) which is of course not the case ?

Comment (sonicaj) [reply src/middlewared/middlewared/plugins/cron.py]: @Mrt134 ping

Comment (sonicaj) [reply src/middlewared/middlewared/plugins/cron.py]: @freenas/docs-team ping

Comment (Mrt134) [reply src/middlewared/middlewared/plugins/cron.py]: There's a couple different ways. It's currently suggested as an understood implication that first item correlates to first item, but you can add `, respectively` to the end of the sentence to clarify or rewrite to break up the paired pairs:
"Any newline or "%" characters in the command are automatically removed or escaped with backslash, respectively."
or
"Newline characters are automatically removed and "%" characters escaped with a backslash."

Comment (sonicaj) [reply src/middlewared/middlewared/plugins/cron.py]: Perfect, thank you :) I'll have this covered.

Comment (Mrt134) [reply src/middlewared/middlewared/plugins/cron.py]: No problem! Sorry it took a bit to get back to you.

**Pull request merged**